crypto/x509/x509_vfy.c

```diff
@@ -519,12 +520,72 @@ static int check_chain_extensions(X509_STORE_CTX *ctx)
                 ret = 1;
             break;
         }
-        if ((x->ex_flags & EXFLAG_CA) == 0
-            && x->ex_pathlen != -1
-            && (ctx->param->flags & X509_V_FLAG_X509_STRICT)) {
-            ctx->error = X509_V_ERR_INVALID_EXTENSION;
-            ret = 0;
+        /*
+         * Do the following set of checks only if strict checking is requrested
+         * and not for self-issued (including self-signed) EE (non-CA) certs
+         * because RFC 5280 does not apply to them according RFC 6818 section 2.
+         */
+        if ((ctx->param->flags & X509_V_FLAG_X509_STRICT) != 0
+            && !(i == 0 && (x->ex_flags & EXFLAG_CA) == 0
+                 && (x->ex_flags & EXFLAG_SI) != 0)) {
```

Comment:
I don't think this is the check we want to do. I think we need to look at num > 1 instead.

Reply:
Using
```
     && !(i == 0 && (x->ex_flags & EXFLAG_CA) == 0
                 && (x->ex_flags & EXFLAG_SI) != 0))
```
I tried to follow https://tools.ietf.org/html/rfc6818#section-2 very closely, which says:

> use of self-issued certificates and self-signed certificates
> issued by entities other than CAs are outside the scope of this specification

Using instead `num > 1` is rather indirect in my view, but seems to have pretty much the same effect in practice.
So I've changed the condition this way.